File: atcoder.jp/contests/hitachi2020/tasks/hitachi2020_a/main.cpp
#include <bits/stdc++.h>
using namespace std;

bool solve(const string &s) {
  if (s == "hi") return true;
  return s.substr(0, 2) == "hi" && solve(s.substr(2));
}

int main() {
  string S;
  cin >> S;

  cout << (solve(S) ? "Yes" : "No") << endl;
}

File: atcoder.jp/contests/hitachi2020/tasks/hitachi2020_a/main.hs
yesNo :: Bool -> String
yesNo False = "No"
yesNo True = "Yes"

solve :: String -> Bool
solve [] = False
solve [_] = False
solve [x, y] = [x, y] == "hi"
solve (x : y : xs) = [x, y] == "hi" && solve xs

main :: IO ()
main = do
  getLine >>= putStrLn . yesNo . solve
